core/src/main/scala/filodb.core/metadata/Schemas.scala

```diff
@@ -0,0 +1,237 @@
+ * sample and is encoded.  It has a unique hash code for each unique DataSchema.
+ * One Dataset in FiloDB can comprise multiple DataSchemas.
+ * One DataSchema should be used for each type of metric or data, such as gauge, histogram, etc.
+ * The "timestamp" or rowkey is the first column and must be either a LongColumn or TimestampColumn.
+ * DataSchemas are intended to be built from config through Schemas.
+ */
+final case class DataSchema private(name: String,
+                                    columns: Seq[Column],
+                                    downsamplers: Seq[ChunkDownsampler],
+                                    hash: Int,
+                                    valueColumn: ColumnId) {
+  val timestampColumn  = columns.head
+  val timestampColID   = 0
+
+  // Used to create a `VectorDataReader` of correct type for a given data column ID;  type PtrToDataReader
+  val readers          = columns.map(col => BinaryVector.defaultPtrToReader(col.columnType.clazz)).toArray
+
+  // The number of bytes of chunkset metadata including vector pointers in memory
+  val chunkSetInfoSize = ChunkSetInfo.chunkSetInfoSize(columns.length)
+  val blockMetaSize    = chunkSetInfoSize + 4
+
+  def valueColName: String = columns(valueColumn).name
+}
+
+/**
+ * A PartitionSchema is the schema describing the unique "key" of each time series, such as labels.
+ * The columns inside PartitionSchema are used for distribution and sharding, as well as filtering and searching
+ * for time series during querying.
+ * There should only be ONE PartitionSchema across the entire Database.
+ */
+final case class PartitionSchema(columns: Seq[Column],
+                                 predefinedKeys: Seq[String],
+                                 options: DatasetOptions) {
+  val binSchema = new RecordSchema(columns.map(c => ColumnInfo(c.name, c.columnType)), Some(0), predefinedKeys)
+
+  private val partKeyBuilder = new RecordBuilder(MemFactory.onHeapFactory, binSchema, 10240)
+
+  /**
+   * Creates a PartitionKey (BinaryRecord v2) from individual parts.  Horribly slow, use for testing only.
+   */
+  def partKey(parts: Any*): Array[Byte] = {
+    val offset = partKeyBuilder.addFromObjects(parts: _*)
+    val bytes = binSchema.asByteArray(partKeyBuilder.allContainers.head.base, offset)
+    partKeyBuilder.reset()
+    bytes
+  }
+}
+
+object DataSchema {
+  import Dataset._
+
+  def validateValueColumn(dataColumns: Seq[Column], valueColName: String): ColumnId Or BadSchema = {
+    val index = dataColumns.indexWhere(_.name == valueColName)
+    if (index < 0) Bad(BadColumnName(valueColName, s"$valueColName not a valid data column"))
+    else           Good(index)
+  }
+
+  /**
+   * Generates a unique 16-bit hash from the column names and types.  Sensitive to order.
+   */
+  def genHash(columns: Seq[Column]): Int = {
+    var hash = 7
+    for { col <- columns } {
+      // Use XXHash to get high quality hash for column name.  String.hashCode is _horrible_
```

Comment:
The default string hashcode is actually quite good, but it's not very efficient, and it doesn't fill the upper bits when hashing short strings. Considering that you're tossing away the upper bits anyhow, it doesn't matter. When calling getBytes always specify a character encoding, or else you'll potentially get different results on different machines.

Reply:
I actually tested it for our specific purposes, and found it to have a terrible distribution.  Not even close to the better algorithms.  The upper bits are actually the most important bits we use for other things, but in this case maybe it doesn't matter.  Anyhow, this will only be used at schema creation time, which is once on load, and good distribution is much more important, so I'd prefer better distribution over efficiency :)